slack/models.py

```diff
@@ -0,0 +1,142 @@
+    # oldest_message_ts (null meaning no limit)
+    oldest_message_ts = models.CharField(max_length=32, null=True)
+    # but not strictly newer than newest_message_ts (null meaning no limit)
+    newest_message_ts = models.CharField(max_length=32, null=True)
+
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.lookups.GreaterThan(
+                    models.functions.Cast(
+                        "newest_message_ts", output_field=models.FloatField()
+                    ),
+                    models.functions.Cast(
+                        "oldest_message_ts", output_field=models.FloatField()
+                    ),
+                ),
+                name="newest_newer_than_oldest",
+            )
+        ]
+
+    def __str__(self):
+        return "(%s, %s)" % (self.oldest_message_ts, self.newest_message_ts)
+
+
+class SeenMessage(models.Model):
+    """
+    DEBUG ONLY: Store all seen messages to double-check we don't see them
+    twice.
+
+    Invariant:
+
+        SeenMessage.objects.count() == \
+            SlackActivityBucket.objects.aggregate(sum=Sum('count'))['sum']
+    """
+
+    channel = models.ForeignKey(Channel, on_delete=models.CASCADE)
+    ts = models.CharField(max_length=32)
+    thread = models.ForeignKey("Thread", on_delete=models.CASCADE, null=True)
+    db_created_at = models.DateTimeField(auto_now_add=True)
+
+    class Meta:
+        unique_together = [("channel", "ts")]
+
+
+class Thread(models.Model):
+    channel = models.ForeignKey(Channel, on_delete=models.CASCADE)
+    thread_ts = models.CharField(max_length=32)
+    last_update_ts = models.CharField(max_length=32, null=True)
+    db_created_at = models.DateTimeField(auto_now_add=True)
+
+    class Meta:
+        unique_together = [("channel", "thread_ts")]
+        constraints = [
+            models.CheckConstraint(
+                check=models.lookups.GreaterThanOrEqual(
+                    models.functions.Cast(
+                        "last_update_ts", output_field=models.FloatField()
+                    ),
+                    models.functions.Cast(
+                        "thread_ts", output_field=models.FloatField()
+                    ),
+                ),
+                name="update_newer_than_created",
+            )
+        ]
+
+
+class SlackActivityBucket(models.Model):
+    """
+    Message count per user per channel per UTC day.
+    """
+
+    day = models.DateField()
+    user = models.ForeignKey(SlackUser, on_delete=models.CASCADE)
+    channel = models.ForeignKey(Channel, on_delete=models.CASCADE)
+    count = models.PositiveIntegerField()
+
+    class Meta:
+        unique_together = [("channel", "day", "user")]
+
+    @classmethod
+    def track_activity(self, channel, user, ts):
+        day = parse_ts(ts).date()
+
+        with connection.cursor() as cursor:
+            # BBB I assume in Django 5.0 can do this with
+            # update_or_create(create_defaults)
```

Comment:
Oh I didn't realize we weren't on Django 5.x. We should upgrade... I'll create a ticket.

Reply:
Sounds good. Maybe replace it with a comment stating this is a performance optimization over using update_or_create.